Sources/EpoxyCollectionView/CollectionView/ReusableViews/CollectionViewCell.swift

```diff
@@ -120,22 +120,31 @@ public final class CollectionViewCell: UICollectionViewCell, ItemCellView {
     // On iOS 15, cells that return an unstable size can result in a layout recursion crash. The
     // root cause of these crashes is likely an ambiguous layout due to misconfigured constraints.
     // Unfortunately, finding each and every one of these ambiguous layouts can be challenging. This
-    // code works around this new behavior / crash by giving up on self-sizing after 5 attempts.
-    // Once a component has been sized differently 5 times, we assume that it has an ambiguous
-    // layout and we simply return the last computed size for the remainder of the component's life.
-    // We reset the count on cell reuse and when the cell is reconfigured in `CollectionView`'s
-    // `configure` function.
-    if CollectionViewConfiguration.shared.enableLayoutRecursionWorkaround {
-      if let previousComputedSize = previousComputedSize, numberOfNewComputedSizes >= 5 {
-        EpoxyLogger.shared.assertionFailure(
-          "Layout recursion detected. View: \(view?.description ?? "nil view"). Size: \(preferredAttributes.size).")
-        preferredAttributes.size = previousComputedSize
+    // code tries to detect these layout issues and logs a warning when one is detected: if a
+    // component size is unstable for at least 50 sizing attempts in less than 2s, we assert since
+    // collection view is about to crash anyways.
+
+    let sizingAttemptTime = CACurrentMediaTime()
+    let wasFirstSizingAttemptRecent: Bool
+    if let firstSizingAttemptTime = firstSizingAttemptTime {
+      wasFirstSizingAttemptRecent = sizingAttemptTime - firstSizingAttemptTime < 2
+      if !wasFirstSizingAttemptRecent {
+        // If the last sizing attempt was not nil, but also more than the 2s threshold, reset all
+        // self-sizing-layout-recursion-tracking state.
+        resetSelfSizingLayoutRecursionTrackingState()
       }
+    } else {
+      firstSizingAttemptTime = sizingAttemptTime
+      wasFirstSizingAttemptRecent = false
+    }
 
-      if preferredAttributes.size != previousComputedSize {
-        numberOfNewComputedSizes += 1
-        previousComputedSize = preferredAttributes.size
-      }
+    if wasFirstSizingAttemptRecent, previousComputedSizes.count >= 50 {
+      EpoxyLogger.shared.assert(
+        "Layout recursion detected. View: \(view?.description ?? "nil view"). Sizes: \(previousComputedSizes).")
```

Comment:
do we want to print the hierarchy this view is in to help us debug which screens / sections are causing this? Here's some sample code:

```swift
extension UIView {
  public var invertedHierarchyString: String {
    var viewDescriptions = [description]
    var next = superview
    while next != nil {
      if let desc = next?.description {
        viewDescriptions.append(desc)
      }
      next = next?.superview
    }
    return hierarchyString(from: viewDescriptions.reversed())
  }

  private func hierarchyString(from viewDescriptions: [String]) -> String {
    var result = ""
    for (idx, desc) in viewDescriptions.enumerated() {
      var prefix = ""
      if idx > 0 {
        prefix = String(repeating: " ", count: 4 * idx) + "├─ "
      }
      result += "\(prefix)\(desc)\n"
    }
    return result
  }
}
```

This would print a view's hierarchy from top to bottom with the last element being the view this was called on

e.g.:
```
<UIView: 0x7f83d782ceb0; frame = (0 0; 390 844); autoresize = W+H; layer = <CALayer: 0x6000026d22e0>>
    ├─ <DLSElementsCoreUI.Button: 0x7f83d782d790; baseClass = UIButton; frame = (0 0; 0 0); opaque = NO; layer = <CALayer: 0x6000026d2520>>
        ├─ <ColorView: 0x7f83d782ee40; frame = (0 0; 0 0); userInteractionEnabled = NO; layer = <CALayer: 0x600002701ca0>>
```

Reply:
Works for me, will add this in!

Reply:
<img width="3002" alt="image" src="https://user-images.githubusercontent.com/746571/156482548-5b1d9dd9-a1fe-4559-8f67-9b1c7847a18f.png">

Happy to keep this in but not sure how useful it is since it only prints view, not VCs. If you look at this screenshot, you can see that there's no reference to `CardStackViewController` or `CollectionViewController` even though that's where the crash is taking place.

Reply:
Ahh that’s a good point @bryankeller. We can’t get the VC from a subview right? I think we can just remove this then